ir/src/main/scala/org/scalajs/ir/Definitions.scala

```diff
@@ -89,190 +451,234 @@ object Definitions {
+  // --------------------------------------------------------------------------
+
+  // --- Validation of encoded names ---
+
+  private def throwInvalidEncodedName(encoded: EncodedName): Nothing = {
+    throw new IllegalArgumentException(
+        "Invalid encoded name: " + new String(encoded, UTF_8))
   }
 
-  private val compressedClasses: Map[String, ClassName] = Map(
-      "java_lang_Object" -> ObjectClass,
-      "java_lang_String" -> BoxedStringClass
-  ) ++ (
-      for (index <- 2 to 22)
-        yield s"scala_Tuple$index" -> ClassName("T" + index)
-  ) ++ (
-      for (index <- 0 to 22)
-        yield s"scala_Function$index" -> ClassName("F" + index)
-  )
+  private def validateSimpleEncodedName(encoded: EncodedName): EncodedName =
+    validateSimpleEncodedName(encoded, 0, encoded.length, openAngleBracketOK = true)
+
+  private def validateSimpleEncodedName(encoded: EncodedName, start: Int,
+      end: Int, openAngleBracketOK: Boolean): EncodedName = {
+
+    if (start == end)
+      throwInvalidEncodedName(encoded)
+    var i = start
+    while (i != end) {
+      val b = encoded(i).toInt
+      if (b >= 0) {
+        // single-byte code point, ASCII repertoire
+        (b: @switch) match {
+          case '.' | ';' | '[' | '/' =>
+            throwInvalidEncodedName(encoded)
+          case '<' =>
+            if (!openAngleBracketOK)
+              throwInvalidEncodedName(encoded)
+          case _ =>
+        }
+        i += 1
+      } else {
+        // multi-byte code point
+        i += validateMultibyteCodePointAndGetByteLen(encoded, end, i, b)
+      }
+    }
 
-  private val decompressedClasses: Map[ClassName, String] =
-    compressedClasses map { case (a, b) => (b, a) }
-
-  private val compressedPrefixes = Seq(
-      "scala_scalajs_runtime_" -> "sjsr_",
-      "scala_scalajs_" -> "sjs_",
-      "scala_collection_immutable_" -> "sci_",
-      "scala_collection_mutable_" -> "scm_",
-      "scala_collection_generic_" -> "scg_",
-      "scala_collection_" -> "sc_",
-      "scala_runtime_" -> "sr_",
-      "scala_" -> "s_",
-      "java_lang_" -> "jl_",
-      "java_util_" -> "ju_"
-  )
+    encoded
+  }
 
-  private val decompressedPrefixes: Seq[(String, String)] =
-    compressedPrefixes map { case (a, b) => (b, a) }
+  private def validateEncodedClassName(encoded: EncodedName): EncodedName = {
+    val len = encoded.length
+    var i = 0
+    while (i < len) {
+      val start = i
+      while (i != len && encoded(i) != '.')
+        i += 1
+      validateSimpleEncodedName(encoded, start, i, openAngleBracketOK = true)
+      i += 1 // goes to `len + 1` iff we successfully parsed the last segment
+    }
 
-  /** Encodes a method name from its full signature. */
-  def encodeMethodName(baseName: String, paramTypes: List[TypeRef],
-      resultType: Option[TypeRef]): MethodName = {
+    if (i == len) // the last segment was empty (including when `len == 0`)
+      throwInvalidEncodedName(encoded)
```

Comment:
This only triggers when `len == 0` right? Otherwise the last call to `validateSimpleEncodedName` would fail. Use a `do ... while`?

```scala
var len = encoded.length
var i = 0
do {
  val start = i
  while (i != len && encoded(i) != '.')
    i += 1
  validateSimpleEncodedName(...)
  i += 1
} while (i < len)
```

Reply:
No it also happens if the last byte of `encoded` is a `'.'`, which is why the comment says "when the last segment is empty". I have expanded the comment to make that clearer.

Reply:
Ah, right. All this manual indexing is hard to parse :P